### DISCUSSION TITLE
Injecting git metadata (tag/hash) into deployments

### DISCUSSION THREAD

**Discussion by fmilichovsky** [fluxcd/flux2#2861] [Q&A]
Hey there,
I'm wondering if flux has any way to use git metadata as variables that could be injected into manifests.
The use-case I'm thinking about is injecting the current git branch/hash/tag of the entire infrastructural repo into a deployment manifest e.g. as an environment variable.

I'd like to avoid manually hardcoding the current "build" (git tag) directly in the code if possible.

**Repo structure**:
- Multiple repositories, a separate one for each service/component, each pushes tagged images as part of its CI.
- Infrastructural repo watched by flux. This one's CI keeps collecting the tags and substitutes them into each component's respective manifest. This conceptually leads to "snapshots" of arbitrary configurations of all the different components in the cluster.

**Intended use-case(s)**:
- For issue reporting/clarity, we want to be able to propagate the current build number/git hash/git tag of the "whole snapshot", i.e. the infrastructural repo, which would make it simple to trace back to the configuration of the whole cluster at that time, including individual versions of all the components.
- This should have no "functional" effect on the cluster itself, its setup is perfectly valid without this piece of "metadata". I would essentially be happy with having easy, clean access to whatever values end up being written to flux's `gitrepository`/`kustomization` resources (the currently applied git hash) as it watches the git repo.

**Comment by kingdonb** [comment]
I proposed this same idea once a long time ago myself and it was rejected by every Flux maintainer (and I felt so strongly about the idea that I actually brought it around to every active Flux maintainer at the time LOL), so I'll take this question!

> I'd like to avoid manually hardcoding the current "build" (git tag) directly in the code if possible.

This is still possible, it requires some compromises today and doesn't work the way you suggested. My suggestion is to maintain the commit hash in a configmap which is updated by your CI process (either through git or directly in the cluster) – what you're doing is to make CI-driven delivery since everything is based on the commit hash and the successful result of CI from that commit hash, you can have a CI process that writes that commit hash to the configmap, commits and pushes it, the delivery can still be all done by Flux and you get most of what you wanted.

There are several examples of this here, but your implementation may vary: [GitHub Actions Manifest Generation](https://fluxcd.io/docs/use-cases/gh-actions-manifest-generation/#manifest-generation-examples)

Then with that configmap you can use [variable substition](https://fluxcd.io/docs/components/kustomize/kustomization/#variable-substitution) and perhaps even trigger Flux from CI to reconcile after the build succeeds and the commit hash is placed in the config map, Flux will replace the commit hash anywhere that it should go in the YAML manifests throughout the system.

The model that you proposed has a major drawback, it only can work effectively with the monorepo approach which doesn't scale and isn't our first choice for Flux users. The approach of writing configmaps into Git from CI by contrast can scale to many repositories.

We recommend to use more repositories as Flux is built to work with many Git repositories; generally each microservice or separate application that you might develop or obtain from a vendor has its own git repository or other type of repo where manifests are hosted. You can also make a model where each environment goes into its own separate repository, so with Flux it's really important to support multiple repositories, I'll stress that we have to support more than just monorepo, one reason why we do not support this feature.

**Reply by kingdonb** [reply]
I didn't want to make my answer too big, (but like I said I went around to every Flux maintainer...) there are more reasons why this feature was not implementable.

When you have more repos, or even when you have a separate bootstrap repo (separate from your app) the local commit hash from the repo that your YAML manifests come from is no longer the correct commit hash to use, so this commit hash is only useful in a very narrow case. This is the first reason, rather than philosophical reasons I think, why this feature idea was rejected by all maintainers.

On the philosophy of GitOps side, there are stronger arguments than that one though.

Mixing the state of the git repository with the identity of the commit changes the GitOps model completely on its head. So it was argued, this mixing is not a decision that Flux should take or suggest for its users, or even offer as an option (following the rationale that if it's a good feature then it should be enabled by default, or always-on, rather than propagating many flags that all change the behavior or isolate users from different features creating new edge cases...)

I'm not sure it's good to allow the working or failing state of the cluster and its workloads to be governed by the identity of the commit. The system should be completely reproducible solely from the set of artifacts that compose it, and even in the absence of a CI system to actualize the new state with an updated commit hash, that seems like it would only make CD wholly dependent on CI again, in a way which delivery should not depend. It should always be possible to take the manifests in the main bootstrap git repo, apply them through a Kustomization, and restore the cluster to working state again with declarative artifacts, even when CI is not available for some reason it should remain possible to update the cluster with Flux, so long as the image repository remains available (and no image references are need to have changed, or artifacts or other repos which have gone missing.)

Rebasing a commit and changing the hash should never cause its manifest content to become invalid, which is what it could do if we facilitated this behavior directly in Flux as you suggest.

In GitOps it should always be possible to restore your system to its working state without running CI again, the commit hash should not function as a magic key that makes the configuration valid, without which it will not function. The whole system should work with any dependency on the commit hash (unless it is kept as data within the system, it can be provided out-of-band from Git) so long as the declarative artifacts are all maintained.

This feature idea (which we both had) breaks hermeticity of the git commit.

**Reply by fmilichovsky** [reply]
Thanks a lot for the replies, I'll be slowly making my way through the docs you've sent to see if I could leverage them, but maybe I should also elaborate more on the structure we have and the use-case I'm aiming for, to make sure we're on the same page on that front.

**Repo structure**:
- Multiple repositories, a separate one for each service/component, each pushes tagged images as part of its CI.
- Infrastructural repo watched by flux. This one's CI keeps collecting the tags and substitutes them into each component's respective manifest. This conceptually leads to "snapshots" of arbitrary configurations of all the different components in the cluster.

**Intended use-case(s)**:
- For issue reporting/clarity, we want to be able to propagate the current build number/git hash/git tag of the "whole snapshot", i.e. the infrastructural repo, which would make it simple to trace back to the configuration of the whole cluster at that time, including individual versions of all the components.

> I'm not sure it's good to allow the working or failing state of the cluster and its workloads to be governed by the identity of the commit. The system should be completely reproducible solely from the set of artifacts that compose it

I might be misunderstanding this bit, what I'm aiming for should have little/no "functional" effect on the cluster itself, its setup is perfectly valid without this piece of "metadata". I would essentially be happy with having easy, clean access to whatever values end up being written to flux's `gitrepository`/`kustomization` resources (the currently applied git hash) as it watches the git repo.

Not sure if this adds any helpful new context, I wasn't entirely sure if we're thinking about the same goal from some of the comments about using this info for determining validity of the system...

**Reply by kingdonb** [reply]
We definitely had a different goal in mind. My use case idea was mostly this, to grab the image reference for use in a dev environment:

```
image: myorg/myapp:${GIT_SHA}
```

There are certainly less problematic uses for this metadata, but no matter which way you're using it, it is the same provision so the same decision either way for Flux maintainers. Using it in this way I propose would break GitOps. Using it in the way you propose definitely improves traceability. I wonder if there's some middle-ground version of the feature idea which does not break the important hermetic property of GitOps deployment. 🤔

**Comment by kingdonb** [comment]
Another possibility which wasn't yet viable when you asked this, but is now in a pre-release state in the `fluxcd/flux2` `oci` branch:

* https://github.com/fluxcd/flux2/pull/2856

If the idea of CI pushing to a Git Repository turns you off, this should sound like a good alternative! CI is already pushing to the image repository, so this does not create any new trust or violate any trust boundary. This has always been a fundamental flaw of Flux's "ImageUpdateAutomation" and even CI-based manifest generation approaches we have often suggested in the past – that writing to Git and maintaining write access to the Git repo from which Flux's workloads are deployed, represents a very grave hazard in that compromising that CI-hosted credential would grant access to rewrite any part of the infrastructure that Flux manages on the cluster.

But the CI is already pushing artifacts to the Image repository, and we can avoid that hazard and all the negative connotations that come with it. We already likely have a mechanism for turning Git commits into container images securely, and we already have trusted that what CI writes out to the registry is safe to run on our clusters. We're already pulling images into the cluster at runtime. The OCI repository can act as a proxy for the Git repository now and host the manifests as well.

You can use `flux push` in a CI process to publish that configmap with the git commit in it, that way it can scale to many repos also. Then the configmap (containing the commit hash) does not need to go into any Git Repository, CI does not have to write to git at all, and you do not need to add another Git repository to your operation or to your cluster. This is really better, but it's in pre-release and based on RFC-0003, which has been posted for review for some time.

We're very excited about this feature and it should be available to use in an actual Flux2 release, if my estimates hold true, by around this time next month. (If you want to get your hands on it before then, you can always build the pre-release `flux` CLI for yourself, and I have an [example CI workflow](https://github.com/kingdonb/podinfo/blob/master/.github/workflows/build-flux-oci-prerelease.yaml) that builds the `flux` CLI from that `oci` branch, so I can test this feature before it has been fully released in my own CI workflows.)
